tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -286,5 +284,103 @@ StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
+  using match::Clamp;
+  using match::Convert;
+  using match::GetTupleElement;
+  using match::Op;
+
+  // The pattern we want to match:
+  //   convert(clamp(broadcast(-128), (get_tuple_element(custom_call(x,w, ...)), broadcast(127));
+  //
+  // With its variants involving commute/reassociation of adds, multiplies, and
+  // max, and omission of alpha1, side_input, alpha2, or bias.
+
+  HloInstruction* gte = nullptr;
+  HloInstruction* conv_instr = nullptr;
+  auto lower_pattern = Broadcast(match::ConstantScalar(-128));
+  auto upper_pattern = Broadcast(match::ConstantScalar(127));
+  auto pattern = Convert(
+      Clamp(lower_pattern,
+            GetTupleElement(
+                &gte, Op(&conv_instr).WithOpcode(HloOpcode::kCustomCall), 0),
+            upper_pattern));
+
+  if (Match(instr, pattern)) {
+    if (primitive_util::IsIntegralType(
+            conv_instr->operand(0)->shape().element_type()) &&
+        instr->shape().element_type() == xla::S8) {
+      HloCustomCallInstruction* conv = CastOrNull<HloCustomCallInstruction>(conv_instr);
+      return ConvWithConvert{instr, gte, conv};
+    }
+  }
+  return absl::nullopt;
+}
+
+Status TryRewriteToCudnnForwardConvert(
+    ConvWithConvert match) {
+  auto conv = match.conv;
+  auto gte = match.gte;
+  auto convert = match.convert;
+
+  // Change type on conv and gte
+  auto convert_out_type = convert->shape().element_type();
+  conv->mutable_shape()->mutable_tuple_shapes(0)->set_element_type(
+      convert_out_type);
+  gte->mutable_shape()->set_element_type(convert_out_type);
+
+  // Remove clamp/convert and so on and just keep get_tuple_element(custom_call(x,w, ...))
+  convert->ReplaceAllUsesWithDifferentShape(gte);
+  conv->parent()->RemoveInstructionAndUnusedOperands(convert);
+
+  return Status::OK();
+}
+
+StatusOr<bool> RunFuseClamp(HloModule* module) {
+  bool changed = false;
+  for (HloComputation* computation : module->MakeNonfusionComputations()) {
+    std::vector<ConvWithConvert> matches;
+    int num_forward_convs = 0;
+    for (auto instr : computation->instructions()) {
+      auto match = FindConvWithConvert(instr);
+      if (match.has_value()) {
+        matches.push_back(*match);
+      }
+      if (auto call = DynCast<HloCustomCallInstruction>(instr)) {
+        if (call->custom_call_target() == kCudnnConvForwardCallTarget) {
+          num_forward_convs++;
+        }
+      }
+    }
+    VLOG(1) << "Identified cuDNN forward conv + convert: " << matches.size()
+            << " out of " << num_forward_convs << " forward convs.";
+    std::vector<std::pair<HloInstruction*, std::unique_ptr<HloInstruction>>>
+        replacements;
+    for (const ConvWithConvert& match : matches) {
+      Status s = TryRewriteToCudnnForwardConvert(match);
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+}  // namespace
+
+StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
+  TF_ASSIGN_OR_RETURN(bool fused_for_bias, RunFuseBiasSideActivation(module));
```

Comment:
I added a second traversal to fused the pattern of custom-call-conv -> clamp -> convert.  It handles both pure convolution and convolution with bias/activation/side cases.

I am open to put the new traversal to a separate transform.